File: android/MatchCraft/app/src/main/java/com/matchcraft/app/keyboard/APIClient.kt
package com.matchcraft.app.keyboard

import com.matchcraft.app.BuildConfig
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Thin wrapper around the v1 Cloudflare Worker's /generate endpoint.
 * Uses OkHttp on a background thread; results are delivered via callback.
 * The Worker URL comes from BuildConfig so it's set per-build in build.gradle.kts.
 */
object APIClient {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

    /**
     * WORKER_BASE_URL is hand-edited in build.gradle.kts, so a typo'd or pasted
     * "http://" value would silently ship screenshots in plaintext. Fail loudly
     * instead. Defence in depth behind network_security_config.xml.
     */
    private val baseUrl: String = BuildConfig.WORKER_BASE_URL.trimEnd('/').also {
        require(it.startsWith("https://")) {
            "WORKER_BASE_URL must use https:// — refusing to send screenshots over plaintext"
        }
    }

    sealed class Result {
        data class Success(val suggestions: List<String>) : Result()
        data class Error(val message: String) : Result()
    }

    fun generate(
        imageBase64: String,
        imageMediaType: String = "image/jpeg",
        tone: String,
        mode: String,
        platform: String? = null,
        onResult: (Result) -> Unit,
    ) {
        val body = JSONObject().apply {
            put("imageBase64", imageBase64)
            put("imageMediaType", imageMediaType)
            put("tone", tone)
            put("mode", mode)
            platform?.let { put("platform", it) }
        }

        val request = Request.Builder()
            .url("$baseUrl/generate")
            .post(body.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                onResult(Result.Error(e.message ?: "Network error"))
            }

            override fun onResponse(call: Call, response: Response) {
                val rawBody = response.body?.string() ?: run {
                    onResult(Result.Error("Empty response"))
                    return
                }

                runCatching {
                    val json = JSONObject(rawBody)

                    if (response.code == 429) {
                        onResult(Result.Error(json.optString("error", "Daily limit reached. Try again tomorrow 💘")))
                        return
                    }

                    val errorMsg = json.optString("error", "")
                    if (errorMsg.isNotEmpty()) {
                        onResult(Result.Error(errorMsg))
                        return
                    }

                    val arr = json.getJSONArray("suggestions")
                    val list = (0 until arr.length()).map { arr.getString(it) }

                    if (list.isEmpty()) {
                        onResult(Result.Error("No suggestions returned"))
                    } else {
                        onResult(Result.Success(list))
                    }
                }.onFailure { e ->
                    onResult(Result.Error("Parse error: ${e.message}"))
                }
            }
        })
    }
}

File: ios/MatchCraft/MatchCraftKeyboard/APIClient.swift
import Foundation

// Update this to your deployed Cloudflare Worker URL (no trailing slash)
let WORKER_BASE_URL = "https://matchcraft-api.YOUR-NAME.workers.dev"

enum APIError: LocalizedError {
    case invalidURL
    case noData
    case serverError(String)
    case decodingError
    case rateLimitExceeded(String)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid API URL — WORKER_BASE_URL in APIClient.swift must be a valid https:// address"
        case .noData: return "No response received"
        case .serverError(let msg): return msg
        case .decodingError: return "Unexpected response format"
        case .rateLimitExceeded(let msg): return msg
        case .networkError(let e): return e.localizedDescription
        }
    }
}

final class APIClient {

    static let shared = APIClient()
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        return URLSession(configuration: config)
    }()

    func generate(
        imageBase64: String,
        imageMediaType: String = "image/jpeg",
        tone: String,
        mode: String,
        platform: String? = nil,
        completion: @escaping (Result<[String], APIError>) -> Void
    ) {
        // WORKER_BASE_URL is hand-edited above, so a typo'd or pasted "http://"
        // value would silently ship screenshots in plaintext. Reject anything
        // that is not HTTPS. Defence in depth behind App Transport Security.
        guard let url = URL(string: "\(WORKER_BASE_URL)/generate"),
              url.scheme?.lowercased() == "https" else {
            completion(.failure(.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = [
            "imageBase64": imageBase64,
            "imageMediaType": imageMediaType,
            "tone": tone,
            "mode": mode,
        ]
        if let platform { body["platform"] = platform }

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            completion(.failure(.networkError(error)))
            return
        }

        session.dataTask(with: request) { data, response, error in
            if let error {
                completion(.failure(.networkError(error)))
                return
            }

            guard let data else {
                completion(.failure(.noData))
                return
            }

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                completion(.failure(.decodingError))
                return
            }

            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 200

            if statusCode == 429, let msg = json["error"] as? String {
                completion(.failure(.rateLimitExceeded(msg)))
                return
            }

            if let errorMsg = json["error"] as? String {
                completion(.failure(.serverError(errorMsg)))
                return
            }

            guard let suggestions = json["suggestions"] as? [String], !suggestions.isEmpty else {
                completion(.failure(.decodingError))
                return
            }

            completion(.success(suggestions))
        }.resume()
    }
}
